### PR TITLE
TS-provided Configurator subtrees

### PR DESCRIPTION
**Pull request by okt-galaktionov**
The minimal set of changes required to allow test suites to register their own Configurator subtrees.

PCH nodes can be registered by ELF constructors added to the Unix Test Agent by linking an external library provided by a test suite.

Testing done: these changes have been successfully used to implement additional TA functionality in one of our test suites

### CODE REVIEW

**Comment by ol-alexandra** [comment]
> Testing done: these changes have been successfully used to implement additional TA functionality in one of our test suites

Please provide some documentation for this new feature.  A link to the TS (if it is public) is also welcome.  It the TS is public, then possibly a link to the patches is just enough.  But we need some documentation in some form.

**Comment by ol-alexandra** [comment]
> The TS isn't public, I'm afraid. I can provide a patch that implements a basic example in selftest, would this work and would it be enough?

Yes, definitely!

File: suites/selftest/conf/builder.conf [hunk 1]
```diff
@@ -65,6 +67,7 @@ define([TE_AGENT_BUILD], [
                          --with-iscsi --with-pcap --with-ppp \
                          --with-socket --with-vxlan])
 
+
             TE_TA_TYPE([${$1_TA_TYPE}], [${$1_TA_TYPE}], [unix],
                        [ --with-8021x --with-aggr --with-iptables \
                          --with-libnetconf --with-ntpd --with-pam \
```
Comment (ol-alexandra) [inline suites/selftest/conf/builder.conf:73]: Unnecessary whitespace change

**Review by ol-alexandra — APPROVED** [review]
Thank you!

File: lib/meson.build [hunk 2]
```diff
@@ -277,9 +279,15 @@ foreach l : all_libs
                 set_variable('te_lib_' + libname, lib)
                 lib_static = lib
 
-                dep = declare_dependency(link_with: lib,
-                                         include_directories: includes,
-                                         dependencies: deps_static)
+                if link_whole
+                    dep = declare_dependency(link_whole: lib,
+                                             include_directories: includes,
+                                             dependencies: deps_static)
```
Comment (ol-andrewr) [inline lib/meson.build:285]: It looks like a dead code in the patch. I see no way to set link_whole==true. Am I missing something?

File: include/te_compiler.h [hunk 3]
```diff
@@ -97,6 +97,18 @@
...
+#if __has_attribute(constructor)
+/**
+ * Mark a function as a constructor. It will be executed
+ * before the main function.
+ */
+#define TE_CONSTRUCTOR_PRIORITY(n) __attribute__((constructor(n)))
+#define TE_CONSTRUCTOR __attribute__((constructor))
+#else
+#define TE_CONSTRUCTOR_PRIORITY(n)
+#define TE_CONSTRUCTOR
```
Comment (ol-andrewr) [inline include/te_compiler.h:109]: It is nice, but I don't understand how it works if we fall into the branch.

File: agents/unix/conf/base/conf_iptables.c [hunk 4]
```diff
@@ -249,7 +249,8 @@ iptables_table_list(unsigned int  gid, const char *oid,
         }
     }
 
-    *list = strdup(table_list.ptr);
+    if (table_list.ptr != NULL)
```
Comment (ol-andrewr) [inline agents/unix/conf/base/conf_iptables.c:252]: As far as I know it is already fixed and will be available.

**Review by ol-andrewr — CHANGES_REQUESTED** [review]

File: agents/unix/conf/base/conf_iptables.c (hunk 4 above)
Comment (ol-andrewr) [reply agents/unix/conf/base/conf_iptables.c:252]: See v1.41.0

File: include/te_compiler.h (hunk 3 above)
Comment (ol-andrewr) [reply include/te_compiler.h:109]: IMHO, it should not result in TE build failure, if the feature is not used.
If these defines are used, it should break the build since it will not work as expected.
The problem is that selftests use it. I guess we can we check it selftests meson and exclude corresponding functionality/components if the feature is not available.

File: agents/unix/meson.build [hunk 5]
```diff
@@ -473,6 +482,7 @@ rcf_listener_sources = files(
 )
 
 # Just use the ta dependencies: this seems to be enough
+# Skip external libraries, however
```
Comment (ol-andrewr) [inline agents/unix/meson.build:485]: What is the real cause of these changes? Sorry, I don't believe that it is just some kind optimization. So, I'd like to know real cause. It must be stated in the changeset description.

In general, the changeset breaks definition of the platform libraries to be linked to all built executables for the platform. The goal of these libraries is to add platform dependencies.

File: agents/unix/main.c [hunk 6]
```diff
@@ -1542,6 +1542,15 @@ rcf_ch_rpc_server_thread(void *ready, int argc, char *argv[])
 }
 #endif
 
+#if TE_CONSTRUCTOR_AVAILABLE
+TE_CONSTRUCTOR_PRIORITY(101)
+#endif
+static void
+log_init(void)
```
Comment (ol-andrewr) [inline agents/unix/main.c:1549]: It looks unused if !TE_CONSTRUCTOR_AVAILABLE. Shouldn't #endif be moved below after the function?

File: suites/selftest/ts/cs/lib/meson.build [hunk 7]
```diff
@@ -0,0 +1,8 @@
+sources += files('main.c')
```
Comment (ol-andrewr) [inline suites/selftest/ts/cs/lib/meson.build:1]: SPDX tag and copyright is missing

File: suites/selftest/ts/cs/lib/main.c [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+{
+    UNUSED(gid);
+    UNUSED(oid);
+
+    strcpy(value, "Hello, world!");
+
+    return 0;
+}
+
+RCF_PCH_CFG_NODE_RO(helloworld, "ts_lib_helloworld", NULL, NULL, helloworld_get);
```
Comment (ol-andrewr) [inline suites/selftest/ts/cs/lib/main.c:34]: It looks unused if constructors are not available

File: agents/unix/meson.build (hunk 5 above)
Comment (ol-andrewr) [reply agents/unix/meson.build:485]: Sorry, explanations are still unclear. What happens if you just remove the patch?

File: agents/unix/meson.build [hunk 9]
```diff
@@ -463,7 +472,7 @@ executable('ta', sources, generated_sources,
            include_directories: includes,
            c_args: c_args,
            link_args: link_args,
-           dependencies: deps)
+           dependencies: deps + ta_external_deps)
```
Comment (ol-andrewr) [inline agents/unix/meson.build:475]: In fact it is a wrong direction.
The key problem here is that when ta_rcf_listener is introduced it breaks TA abstraction from underlying communication protocol. In theory, it could be anything, but rcf_comm_agent_create_listener() hardwires it to TCP over IPv4.

The correct fix is to build ta_rcf_listener if and only if comm_net_agent is present in TA libraries  )agent-unix-libs) and do not add anything except required minimum (logger libs) plus comm_net_agent to ta_rcf_listener dependencies. It does not need any RPC and CS libraries as I understand.

**Review by ol-andrewr — CHANGES_REQUESTED** [review]

**Review by ol-andrewr — APPROVED** [review]

**Review by ol-arteman — COMMENTED** [review]
The component name is just `include`.

File: include/te_compiler.h [hunk 10]
```diff
@@ -97,6 +97,20 @@
 #define TE_DEPRECATED
 #endif
 
+#if __has_attribute(constructor)
```
[the remark this answers is not in the document]
Comment (ol-arteman) [reply include/te_compiler.h:100]: This is still unanswered.

File: agents/unix/main.c [hunk 11]
```diff
@@ -1587,7 +1596,9 @@ main(int argc, char **argv)
 #endif /* AT_SYSINFO */
 #endif
 
-    te_log_init("(unix)", te_log_message_file);
+#if ! TE_CONSTRUCTOR_AVAILABLE
```
[the remark this answers is not in the document]
Comment (ol-arteman) [reply agents/unix/main.c:1599]: Still unanswered.

File: suites/selftest/ts/cs/lib/main.c [hunk 12]
```diff
@@ -0,0 +1,51 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/** @file
+ * @brief TS-provided subtree
+ *
+ * Copyright (c) 2024 OKTET Ltd. All rights reserved.
```
Comment (ol-arteman) [inline suites/selftest/ts/cs/lib/main.c:5]: The copyright in new files should be on the second line, like:
```
/* SPDX-License-Identifier: Apache-2.0 */
/* Copyright (c) 2024 OKTET Ltd. All rights reserved. */
```

**Review by ol-arteman — CHANGES_REQUESTED** [review]
The component is not `selftest:` but rather `suites/selftest:`

**Review by ol-arteman — APPROVED** [review]
